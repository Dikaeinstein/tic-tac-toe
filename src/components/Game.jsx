import React from 'react';
import Board from './Board';
import Title from './Title';
import Footer from './Footer';
import Button from './Button';

import calculateWinner from '../lib/calculateWinner';

class Game extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
    this.handleStart = this.handleStart.bind(this);
    this.state = {
      history: [
        {
          squares: Array(9).fill(null),
          row: 0,
          col: 0,
        },
      ],
      stepNumber: 0,
      xIsNext: true,
      hasStarted: false,
    };
  }

  handleClick(row, col) {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const current = history[history.length - 1];
    const squares = current.squares.slice();
    if (calculateWinner(squares) || squares[col]) {
      return;
    }
    squares[col] = this.state.xIsNext ? 'X' : 'O';
    this.setState({
      history: history.concat([{
        squares,
        col,
        row,
      }]),
      stepNumber: history.length,
      xIsNext: !this.state.xIsNext,
    });
  }

  handleStart() {
    if (!this.state.hasStarted) {
      this.setState({ hasStarted: !this.state.hasStarted });
    } else {
      this.setState({
        history: [
          {
            squares: Array(9).fill(null),
            row: 0,
            col: 0,
          },
        ],
        stepNumber: 0,
      });
    }
  }

  render() {
    const { history } = this.state;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);
    let status;
    let winnerText;

    if (winner) {
      status = `winner is ${winner.winner}`;
      winnerText = 'winner-text';
    } else if (current.squares.indexOf(null) === -1) {
      status = 'Its a Draw!';
      winnerText = 'winner-text';
    } else {
      status = `Next player: ${this.state.xIsNext ? 'X' : 'O'}`;
      winnerText = '';
    }

    return (
      <div>
        <div className="row board-row">
          <div className="col-2 text-center game">
            <div className="title">
              <Title />
            </div>
            <div className="game-board">
              <Board
                squares={current.squares}
                onClick={this.handleClick}
                hasStarted={this.state.hasStarted}
              />
            </div>
            <div className="start-game text-center">
              <Button onClick={this.handleStart} hasStarted={this.state.hasStarted} />
            </div>
          </div>
          <div className="col-1 text-center game-info">
            <div className="player-info">Player1(x)</div>
            <div className="player-info">Player2(0)</div>
            <div className={winnerText}>{status}</div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

export default Game;
