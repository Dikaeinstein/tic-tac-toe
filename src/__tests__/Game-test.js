import React from 'react';
import renderer from 'react-test-renderer';
import Game from '../components/Game';

describe('Game', () => {
  it('Changes Board state when clicked', () => {
    const component = renderer.create(<Game />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
