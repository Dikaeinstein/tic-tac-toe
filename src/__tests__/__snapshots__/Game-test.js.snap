// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Changes Board state when clicked 1`] = `
<div>
  <div
    className="row board-row"
  >
    <div
      className="col-2 text-center game"
    >
      <div
        className="title"
      >
        <h3>
          Tic Tac Toe
        </h3>
      </div>
      <div
        className="game-board"
      >
        <div>
          <div
            className="board-row"
          >
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
          </div>
          <div
            className="board-row"
          >
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
          </div>
          <div
            className="board-row"
          >
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
            <button
              className="square"
              disabled={true}
              onClick={[Function]}
            />
          </div>
        </div>
      </div>
      <div
        className="start-game text-center"
      >
        <button
          className="btn btn-default"
          onClick={[Function]}
        >
          Start Game
        </button>
      </div>
    </div>
    <div
      className="col-1 text-center game-info"
    >
      <div
        className="player-info"
      >
        Player1(x)
      </div>
      <div
        className="player-info"
      >
        Player2(0)
      </div>
      <div
        className=""
      >
        Next player: X
      </div>
    </div>
  </div>
  <footer
    className="text-center"
  >
    <p>
      Made with love and logic by:
      <a
        href="https://github.com/Dikaeinstein"
      >
         Dikaeinstein
      </a>
    </p>
  </footer>
</div>
`;
